bootstrap/src/sun/nio/ch/lincheck/EventTracker.java

```diff
@@ -59,13 +59,16 @@ void cacheInvokeDynamicCallSite(
 
     void updateSnapshotBeforeConstructorCall(Object[] objs);
 
-    void beforeReadField(ThreadDescriptor descriptor, int codeLocation, Object obj, int fieldId);
-    void beforeReadArrayElement(ThreadDescriptor descriptor, int codeLocation, Object array, int index);
+    void beforeReadField(ThreadDescriptor descriptor, int codeLocation, Object obj, int fieldId, String typeDescriptor);
+    void beforeReadArrayElement(ThreadDescriptor descriptor, int codeLocation, Object array, int index, String typeDescriptor);
```

Comment:
There is no need to add `String typeDescriptor` as an additional argument, it can be passed via `FieldDescriptor` class. 

I've already implemented alternative approach in my mirror branch.